cli/azd/pkg/tools/bicep/bicep.go

```diff
@@ -5,170 +5,225 @@ package bicep
 
 import (
 	"context"
+	"errors"
 	"fmt"
+	"io"
 	"log"
+	"net/http"
 	"os"
-	"os/user"
 	"path/filepath"
 	"runtime"
-	"strings"
 
-	"github.com/azure/azure-dev/cli/azd/pkg/convert"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+	"github.com/azure/azure-dev/cli/azd/pkg/config"
 	"github.com/azure/azure-dev/cli/azd/pkg/exec"
+	"github.com/azure/azure-dev/cli/azd/pkg/input"
+	"github.com/azure/azure-dev/cli/azd/pkg/osutil"
 	"github.com/azure/azure-dev/cli/azd/pkg/tools"
 	"github.com/blang/semver/v4"
 )
 
+// cBicepVersion is the minimum version of bicep that we require (and the one we fetch when we fetch bicep on behalf of a
+// user).
+var cBicepVersion semver.Version = semver.MustParse("0.12.40")
+
 type BicepCli interface {
-	tools.ExternalTool
 	Build(ctx context.Context, file string) (string, error)
 }
 
-func NewBicepCli(commandRunner exec.CommandRunner) BicepCli {
-	return &bicepCli{
-		commandRunner: commandRunner,
-	}
+// NewBicepCli creates a new BicepCli. Azd manages its own copy of the bicep CLI, stored in `$AZD_CONFIG_DIR/bin`. If
+// bicep is not present at this location, or if it is present but is older than the minimum supported version, it is
+// downloaded.
+func NewBicepCli(
+	ctx context.Context,
+	console input.Console,
+	commandRunner exec.CommandRunner,
+) (BicepCli, error) {
+	return newBicepCliWithTransporter(ctx, console, commandRunner, http.DefaultClient)
 }
 
-type bicepCli struct {
-	commandPath   *string
-	commandRunner exec.CommandRunner
-}
+// newBicepCliWithTransporter is like NewBicepCli but allows providing a custom transport to use when downloading the
+// bicep CLI, for testing purposes.
+func newBicepCliWithTransporter(
+	ctx context.Context,
+	console input.Console,
+	commandRunner exec.CommandRunner,
+	transporter policy.Transporter,
```

Comment:
I don't feel super strong about re-using the SDK type here as a general HTTP client.
My understanding is that the SDK Transporter is meant specifically for the SDK Pipeline.
I wonder if we would prefer using some Go-http-client interface instead.

Reply:
I don't know of an interface in the standard library that just has the `Do` method, which is what we want, so I always reach for this Transporter interface.

Reply:
Creating our own interface within the bicep package would remove this dependency:
"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"

I think that's a better approach because someone who want to create a bicepCli,  should not depend on the Azure SDK for doing that.

Reply:
I'm going to keep this as `policy.Transporter` for now.  This method is not exported (so external packages couldn't take a dependency on it anyway) and I'd rather not introduce yet another way to spell "an interface which has a single method, Do(*http.Request) (*http.Respose, error)". For cases where the azure SDK already has a way to name the concept, I'd rather we just use it - it helps us speak the same language as the rest of the Gophers on the larger team.